metrics: add a metric for max observed endpoint ifindex

asauber opened the pull request:
Adds a metric `endpoint_max_ifindex`, which is the current maximum
interface index for all endpoints. The metric is updated during the
periodic invocation of `syncHostIPs`.

This metric can be used to determine if the interface index for the next
Pod will exceed Cilium's limit of max(uint16) on older kernels. On
kernels which do not provide ifindex via the FIB, Cilium must store the
ifindex in the CT map, and 16 bits are reserved per entry for this
purpose. This presents a limit of 65535 lifetime interfaces, which can
be approached quickly with sufficient pod churn.

Users who are subject to this limitation (typically a kernel version
less than 5.10), are advised to reboot or roll the host in this case.

Its default-enabled status is dynamic. On kernels which provide ifindex
via the FIB, the metric is disabled (since Cilium's ifindex limits are
not subject to max(uint16) in this case).

It can be enabled with the following Helm values:

```yaml
prometheus:
  enabled: true
  metrics:
    - +cilium_endpoint_max_ifindex
```

On kernels which do not provide ifindex via the FIB, the metric is enabled by default.

### Dynamic default based on Kernel version

On Linux 4.19 without metric explicitly enabled

```
$ cat contrib/testing/kind-values.yaml | grep -A 1 prom
prometheus:
  enabled: true

$ curl 2>/dev/null 172.18.0.4:9962/metrics | grep ifindex
# HELP cilium_endpoint_max_ifindex Maximum interface index observed for existing endpoints
# TYPE cilium_endpoint_max_ifindex gauge
cilium_endpoint_max_ifindex 7

$ uname -r
kind-4.19
```

On Linux 6.4 without metric explicitly enabled

```
$ cat contrib/testing/kind-values.yaml | grep -A 1 prom
prometheus:
  enabled: true
  
$ curl 2>/dev/null 172.19.0.3:9962/metrics | grep ifindex

$ 
```

On Linux 6.4 with metric explicitly enabled

```
$ cat contrib/testing/kind-values.yaml | grep -A 3 prom
prometheus:
  enabled: true
  metrics:
    - +cilium_endpoint_max_ifindex

$ curl 2>/dev/null 172.19.0.3:9962/metrics | grep ifindex
# HELP cilium_endpoint_max_ifindex Maximum interface index observed for existing endpoints
# TYPE cilium_endpoint_max_ifindex gauge
cilium_endpoint_max_ifindex 11

$ uname -r
6.4.12-arch1-1
```

### Longer-running example

Roll a pod or two

```
$ k delete pod -n kube-system          coredns-5d78c9869d-5w8pc
pod "coredns-5d78c9869d-5w8pc" deleted
```

and the `ifindex` increases

```
$ curl 172.19.0.2:9962/metrics 2>/dev/null | grep ifindex
# HELP cilium_endpoint_max_ifindex Maximum interface index observed for existing endpoints
# TYPE cilium_endpoint_max_ifindex gauge
cilium_endpoint_max_ifindex 43
```

Addresses: https://github.com/cilium/cilium/issues/17259
Addresses: https://github.com/cilium/cilium/issues/16260

```release-note
metrics: add a metric for max observed endpoint ifindex
```

asauber committed:
metrics: add metric for max seen ifindex

Add a metric for the current maximum observed interface index

Adds a metric `endpoint_max_ifindex`, which is the current maximum
interface index for all endpoints. The metric is updated during the
periodic invocation of `syncHostIPs`.

This metric can be used to determine if the interface index for the next
Pod will exceed Cilium's limit of max(uint16) on older kernels. On
kernels which do not provide ifindex via the FIB, Cilium must store the
ifindex in the CT map, and 16 bits are reserved per entry for this
purpose. This presents a limit of 65535 lifetime interfaces, which can
be approached quickly with sufficient pod churn.

Users who are subject to this limitation (typically a kernel version
less than 5.10), are advised to reboot or roll the host in this case.

Its default-enabled status is dynamic. On kernels which provide ifindex
via the FIB, the metric is disabled (since Cilium's ifindex limits are
not subject to max(uint16) in this case).

Addresses: https://github.com/cilium/cilium/issues/17259
Addresses: https://github.com/cilium/cilium/issues/16260

Signed-off-by: Andrew Sauber <2046750+asauber@users.noreply.github.com>

asauber committed:
testing: move testutils/ipam into its own package

To avoid a circular import, move testutils/ipam.go into its own package.

The circular import was observed when attempting to use datapath kernel
feature probes in the metrics package. It was affecting only the probe
package's tests.

import cycle:
github.com/cilium/cilium/pkg/datapath/linux/probes
    [github.com/cilium/cilium/pkg/datapath/linux/probes.test]
github.com/cilium/cilium/pkg/testutils
    [github.com/cilium/cilium/pkg/datapath/linux/probes.test]
github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2
    [github.com/cilium/cilium/pkg/datapath/linux/probes.test]
github.com/cilium/cilium/pkg/policy/api
    [github.com/cilium/cilium/pkg/datapath/linux/probes.test]
github.com/cilium/cilium/pkg/metrics
    [github.com/cilium/cilium/pkg/datapath/linux/probes.test]

Moving this package avoids the import cycle

Signed-off-by: Andrew Sauber <andrew.sauber@isovalent.com>

derailed reviewed (approved):
@asauber Nice work! LG. 
For my info: I am not sure what would this new metrics indicates if a system is nearing the max? and what can a user do about it?

christarazi commented on pkg/metrics/metrics.go:
```diff
@@ -570,6 +573,7 @@ type LegacyMetrics struct {
 	NodeConnectivityStatus           metric.Vec[metric.Gauge]
 	NodeConnectivityLatency          metric.Vec[metric.Gauge]
 	Endpoint                         metric.GaugeFunc
+	EndpointMaxIfindex               metric.Gauge
```
Why isn't this `metrics.Vec[metric.Gauge]`? I presume the answer is that we don't have labels for this metric which is fine, but would that eliminate the possibility of using `cluster`, `namespace`, `instance` labels that are added automatically? (at least I think they are).

asauber replied:
I'll confirm whether or not those are available with some local testing.

asauber replied:
Using `metric.Vec` isn't necessary in order to get those additional labels. They come from the prometheus scraper config. As an example, you will find config similar to the following for all metrics endpoints configured using the Prometheus operator:

```yaml
- job_name: serviceMonitor/kube-system/cilium-agent/0
  ...
  scrape_interval: 10s
  metrics_path: /metrics
  - source_labels:
    - __meta_kubernetes_namespace
    target_label: namespace
```

With a Grafana instance hooked up, we can see that many labels are added to this plain `metric.Gauge` with a typical configuration:
 
![Screenshot_20230911_142730-1](https://github.com/cilium/cilium/assets/2046750/e54e135e-17d1-4025-a9fc-61e736904c80)

christarazi commented on Documentation/observability/metrics.rst:
Can you add an upgrade note as well about the newly added metric?

asauber replied:
Yes, I like that idea.

christarazi commented on pkg/metrics/metrics.go:
This commit can be squashed with the previous commit on this file. No need for separate commits that have diffs between them.

christarazi reviewed (changes requested):
Thanks for the PR!

A few comments below. One other thing is that I don't see any commit descriptions for the commits and all the context is instead in the PR. As a project, we are not opposed to a PR description, but we state in our [contrib guide](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request) that PRs should contain commits with a description. Plus, this is beneficial when bisecting / git blame.

asauber committed:
testing: allow linux probes to compile on macOS

For the sake of ease of development, we test in CI that the Ginkgo suite
found in /test can compile on multiple platforms, including darwin.

This test suite includes a test of the Hive framework, which initializes
an instance of the daemon Cell. This Cell requires the MetricsRegistry,
and so these tests depend on the MetricsRegistry. Some metrics may soon
depend upon Linux feature probes, and so make the Linux probes package
an indirect dependency of the Ginkgo tests.

In order to support this build dependency, we allow the probes package
to compile on darwin by adding stub values for netlink constants used by
the package, and a linux-specific declaration of the constants using the
netlink package.

Signed-off-by: Andrew Sauber <2046750+asauber@users.noreply.github.com>

learnitall reviewed (approved):
Looks great on docs! Thank you for adding the description as to the dynamic default status, awesome stuff.

asauber committed:
docs: add upgrade note for max_ifindex metric

Signed-off-by: Andrew Sauber <andrew.sauber@isovalent.com>
